File: recycler_view/recylcer_view1/src/main/java/com/example/recycler_view/MainActivity.java
package com.example.recycler_view;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;


public class MainActivity extends AppCompatActivity {

    ArrayList<Dictionary> mArrayList;
    private CustomAdapter mAdapter;
    private int count = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView mRecyclerView = (RecyclerView) findViewById(R.id.recyclerview_main_list);

        LinearLayoutManager mLinearLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLinearLayoutManager);

        mArrayList = new ArrayList<>();
        mAdapter = new CustomAdapter(mArrayList);
        mRecyclerView.setAdapter(mAdapter);


        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(mRecyclerView.getContext(),
                mLinearLayoutManager.getOrientation());
        mRecyclerView.addItemDecoration(dividerItemDecoration);

        mRecyclerView.addOnItemTouchListener(new RecyclerTouchListener(getApplicationContext(), mRecyclerView, new ClickListener() {

            @Override
            public void onClick(View view, int position) {
                Log.d("onClick😀", "called");
                Dictionary dict = mArrayList.get(position);
                Toast.makeText(getApplicationContext(), dict.getId() + ' ' + dict.getEnglish() + ' ' + dict.getKorean(), Toast.LENGTH_LONG).show();


                Intent intent = new Intent(getBaseContext(), ResultActivity.class);

                intent.putExtra("id", dict.getId());
                intent.putExtra("english", dict.getEnglish());
                intent.putExtra("korean", dict.getKorean());

                startActivity(intent);
            }

            @Override
            public void onLongClick(View view, int position) {
                Log.d("onLongClick😀", "will it be called?");

            }

        }));

        Button buttonInsert = (Button) findViewById(R.id.button_main_insert);
        Button buttonDelete = (Button) findViewById(R.id.button_main_delete);
        buttonInsert.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                count++;

                Dictionary data = new Dictionary(count + "", "Apple" + count, "사과" + count);
                mArrayList.add(data);

                mAdapter.notifyItemInserted(count);

                if (count > 0) {
                    buttonDelete.setVisibility(View.VISIBLE);
                }
            }
        });

        buttonDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                count--;

                if (mArrayList.size() > 0) {
                    mArrayList.remove(mArrayList.size() - 1);
                }
                if (mArrayList.size() <= 0) {
                    buttonDelete.setVisibility(View.GONE);
                    count = 0;
                }
                mAdapter.notifyItemChanged(count);
            }
        });
    }

    public interface ClickListener {
        void onClick(View view, int position);

        void onLongClick(View view, int position);
    }

    public class RecyclerTouchListener implements RecyclerView.OnItemTouchListener {

        private final GestureDetector gestureDetector;
        private final MainActivity.ClickListener clickListener;

        public RecyclerTouchListener(Context context, final RecyclerView recyclerView, final MainActivity.ClickListener clickListener) {
            this.clickListener = clickListener;
            this.gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {

                @Override
                public boolean onSingleTapUp(MotionEvent e) {

                    return true;
                }

                @Override
                public void onLongPress(MotionEvent e) {
                    View child = recyclerView.findChildViewUnder(e.getX(), e.getY());

                    if (child != null && clickListener != null) {
                        clickListener.onLongClick(child, recyclerView.getChildAdapterPosition(child));
                    }

                    Toast toast = Toast.makeText(MainActivity.this, "stop pressing so hard", Toast.LENGTH_LONG);
                    toast.show();
                }
            });
        }

        @Override
        public boolean onInterceptTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {
            View child = rv.findChildViewUnder(e.getX(), e.getY());
            if (child != null && clickListener != null && gestureDetector.onTouchEvent(e)) {
                clickListener.onClick(child, rv.getChildAdapterPosition(child));
            }

            return false;
        }

        @Override
        public void onTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {

        }

        @Override
        public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {

        }
    }
}

File: udp_client_app/app/src/main/java/com/example/udp_client_app/MainActivity.kt
package com.example.udp_client_app

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView

class MainActivity : AppCompatActivity() {
    var udpClientThread: UdpClientThread? = null
    lateinit var udpClientHandler: UdpClientHandler

    lateinit var btnConnect: Button
    lateinit var etAddress: EditText
    lateinit var etPort: EditText
    lateinit var tvState: TextView
    lateinit var tvReceived: TextView

    companion object {
        lateinit var instance: MainActivity
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        instance = this

        btnConnect = findViewById(R.id.btn_connect)
        btnConnect.setOnClickListener(clickListener)

        etAddress = findViewById(R.id.et_address)
        etPort = findViewById(R.id.et_port)
        tvState = findViewById(R.id.tv_state)
        tvReceived = findViewById(R.id.tv_received)

        udpClientHandler = UdpClientHandler()
    }

    private val clickListener = View.OnClickListener {
        val address = if (etAddress.text.isEmpty()) "0.0.0.0" else etAddress.text.toString()
        val port = if (etPort.text.isEmpty()) 0 else Integer.parseInt(etPort.text.toString())

        udpClientThread = UdpClientThread(
            address,
            port,
            udpClientHandler
        )
        udpClientThread?.start()
        btnConnect.isEnabled = false
    }

    fun updateState(msg: String) {
        tvState.text = msg
    }

    fun updateReceivedMsg(msg: String) {
        tvReceived.append("$msg\n")
    }

    fun     clientEnd() {
        udpClientThread = null
        tvState.text = "client end"
        btnConnect.isEnabled = true
    }
}